### PR TITLE
Serialization Perf

### PR DESCRIPTION
**Pull request by rawfalafel**
### Results

Here's the numbers on my computer for the existing benchmark with 150 byte txs
```
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/sharding
BenchmarkSerialization10-12                30000             69008 ns/op           27805 B/op        353 allocs/op
BenchmarkSerialization100-12                5000            739374 ns/op          324040 B/op       3425 allocs/op
BenchmarkSerialization1000-12                300           5558360 ns/op         3127289 B/op      34044 allocs/op
BenchmarkSerialization10000-12                30          48292393 ns/op        63835295 B/op     160091 allocs/op
PASS
```

Here's the numbers for the equivalent benchmark on this branch:

```
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/sharding
BenchmarkSerializeRoundtrip10-12          100000             37949 ns/op           15103 B/op        210 allocs/op
BenchmarkSerializeRoundtrip100-12           3000            426936 ns/op          150057 B/op       2013 allocs/op
BenchmarkSerializeRoundtrip1000-12           500           3309168 ns/op         1483492 B/op      20020 allocs/op
PASS
```

_Note: I took out benchmark tests with 10k transactions because they were hitting the [collation size limit](https://github.com/prysmaticlabs/geth-sharding/blob/master/sharding/collation.go#L150) and failing silently._

The benchmark tests are now broken up into Serialization/Deserialization, and for each of those I included a benchmark of the process without RLP.

```
goos: linux
goarch: amd64
pkg: github.com/ethereum/go-ethereum/sharding
BenchmarkSerializeNoRLP10-12             1000000              3409 ns/op            2088 B/op         11 allocs/op
BenchmarkSerializeNoRLP100-12             100000             34454 ns/op           20880 B/op        101 allocs/op
BenchmarkSerializeNoRLP1000-12              5000            330732 ns/op          200609 B/op       1001 allocs/op
BenchmarkSerialize10-12                   100000             11328 ns/op            4890 B/op         52 allocs/op
BenchmarkSerialize100-12                   10000            131254 ns/op           48999 B/op        502 allocs/op
BenchmarkSerialize1000-12                   1000           1251594 ns/op          481066 B/op       5003 allocs/op
BenchmarkDeserialize10-12                  50000             25462 ns/op           10208 B/op        158 allocs/op
BenchmarkDeserialize100-12                 10000            246955 ns/op          101008 B/op       1511 allocs/op
BenchmarkDeserialize1000-12                  500           3257192 ns/op         1001747 B/op      15014 allocs/op
BenchmarkDeserializeNoRLP10-12            500000              4435 ns/op            2736 B/op         16 allocs/op
BenchmarkDeserializeNoRLP100-12           100000             33656 ns/op           26480 B/op        109 allocs/op
BenchmarkDeserializeNoRLP1000-12            5000            333305 ns/op          257520 B/op       1012 allocs/op
BenchmarkSerializeRoundtrip10-12           50000             44685 ns/op           15103 B/op        210 allocs/op
BenchmarkSerializeRoundtrip100-12           5000            409291 ns/op          150061 B/op       2013 allocs/op
BenchmarkSerializeRoundtrip1000-12           300           5520262 ns/op         1483490 B/op      20019 allocs/op
PASS
```

RLP encoding now consumes 70-80% of wall clock time when running these benchmarks. Optimizing RLP would be the next logical step, but according to the [wiki](https://github.com/ethereum/wiki/wiki/RLP), the geth team are already aware that its inefficient and in the process of addressing perf issues.

### Optimizations
In the existing code, blob serialization/deserialization appends encoded data to an array in a loop. When data's appended to an array at full capacity, go allocates a new array with enough capacity and copies all of the data over. At 10k transactions, this becomes pretty inefficient.

```
var data []byte
for i := 0; i < len(blobs); i++ {
  // this allocates a new array on every iteration!
  data = append(data, byte(i))
}
```

Instead, I created an array with enough capacity
```
data := make([]byte, 0, len(blobs)
for i:= 0; i < len(blobs); i++ {
  data = append(data, byte(i)
}
```

I'm sure there are other optimizations that could be made, but this seemed to be the most glaring one to me. The next optimization would be to avoid encoding/decoding with both RLP and blob serialization.

Edit: Updated with `-benchmem` numbers

### CODE REVIEW

File: sharding/collation_test.go [hunk 1]
```diff
@@ -131,34 +132,153 @@ func makeTxWithGasLimit(gl uint64) *types.Transaction {
 // BENCHMARK TESTS
 
 // Helper function to generate test that completes round trip serialization tests for a specific number of transactions.
-func runBenchTest(b *testing.B, numTransactions int) {
+func makeRandomTransactions(numTransactions int) []*types.Transaction {
 	var txs []*types.Transaction
 	for i := 0; i < numTransactions; i++ {
 		data := make([]byte, 650)
```
Comment (rawfalafel) [inline sharding/collation_test.go]: Tweaking this number affects the benchmark as well. I take it that the average ethereum transaction is around 650 bytes?

File: sharding/utils/marshal_test.go [hunk 2]
```diff
@@ -84,5 +83,127 @@ func TestSerializeAndDeserializeblob(t *testing.T) {
 			t.Errorf("Error Serializing blobs at index %d, the serialized and deserialized versions are not the same:\n\n %v \n\n %v \n\n %v", i, blob, serializedblob, raw)
 		}
 	}
+}
+
+func TestDeserializeSkipEvm(t *testing.T) {
```
Comment (rawfalafel) [inline sharding/utils/marshal_test.go]: The existing tests can cover a lot of cases but I think they could cause some problems down the road. We should avoid testing against objects created with a random number generator because they can fail randomly and in unexpected ways. They also don't test specific cases, meaning that when they do fail, its incredibly unclear what caused the failure. Finally they take a relatively long time to run, and we want to be able to run the entire unit test suite as quickly as possible!

File: sharding/collation_test.go (hunk 1 above)
Comment (nisdas) [reply sharding/collation_test.go]: I would say around 150 bytes per transaction going from the last few blocks in etherscan

File: sharding/utils/marshal_test.go (hunk 2 above)
Comment (nisdas) [reply sharding/utils/marshal_test.go]: Agreed, the reason I used random number generator was to make it easier to uncover edge cases. Currently there is not standardised tests yet for serialisation but in the future there will probably be standardised json tests so we would not be needing to use RNG.

**Comment by nisdas** [comment]
Hey thanks for this ! Really like how you have changed optimisations you have done in both the `Serialize` and `Deserialize` methods.  Everything is more readable and modular.  Great Job !

**Comment by terenc3t** [comment]
There have been conversations in [ethereum/sharding](https://gitter.im/ethereum/sharding?utm_source=share-link&utm_medium=link&utm_campaign=share-link) Gitter channel to use protobuf. I interested to see what performance gain we get instead of rlp. 70-80% is quite a lot.

File: sharding/collation_test.go (hunk 1 above)
Comment (prestonvanloon) [reply sharding/collation_test.go]: There is no basis for the number 650. I picked it at random.

**Comment by prestonvanloon** [comment]
This is awesome! Is this ready for a full review?

**Comment by rawfalafel** [comment]
the build seems to be failing on travis but otherwise, yes! i'll take a look at it tomorrow morning (I'm in Japan time).

**Comment by terenc3t** [comment]
travis timed out, I just restarted again

File: sharding/utils/marshal_test.go [hunk 3]
```diff
@@ -84,5 +83,127 @@ func TestSerializeAndDeserializeblob(t *testing.T) {
...
+	if err != nil {
+		t.Errorf("Serialize failed: %v", err)
+	}
+
+	if len(data) != 64 {
+		t.Errorf("Length of serialized data incorrect. Should be %v but is %v", 32, len(data))
+	}
+
+	if data[32] != 29 {
+		t.Errorf("Indicating byte for second chunk should be %v but is %v", 29, data[0])
```
Comment (prestonvanloon) [inline sharding/utils/marshal_test.go]: Is this correct, should it be `data[32]`?

File: sharding/utils/marshal_test.go [hunk 4]
```diff
@@ -84,5 +83,127 @@ func TestSerializeAndDeserializeblob(t *testing.T) {
...
+		blobData[i] = byte(i)
+	}
+
+	data, err := Serialize(rawBlobs)
+	if err != nil {
+		t.Errorf("Serialize failed: %v", err)
+	}
+
+	if len(data) != 64 {
+		t.Errorf("Length of serialized data incorrect. Should be %v but is %v", 32, len(data))
```
Comment (prestonvanloon) [inline sharding/utils/marshal_test.go]: should be 64?

File: sharding/utils/marshal.go [hunk 5]
```diff
@@ -45,149 +48,140 @@ func ConvertFromRawBlob(blob *RawBlob, i interface{}) error {
...
+	chunksNumber := len(data) / int(chunkSize)
+	serializedBlobs := []SerializedBlob{}
+	numPartitions := 0
+
+	// first iterate through every chunk and identify blobs and their length
+	for i := 0; i < chunksNumber; i++ {
+		indicatorIndex := i * int(chunkSize)
+		databyteLength := getDatabyteLength(data[indicatorIndex])
+
+		// if indicator is non-terminal, increase blobSize by 31
```
Comment (prestonvanloon) [inline sharding/utils/marshal.go]: What does this comment mean? I don't see any increased blobSize by 31?

File: sharding/utils/marshal.go [hunk 6]
```diff
@@ -45,149 +48,140 @@ func ConvertFromRawBlob(blob *RawBlob, i interface{}) error {
...
-	//Loops through all the blobs and serializes them into chunks
-	for i := int64(0); i < length; i++ {
-		data := *rawblobs[i]
-		refinedData, err := SerializeBlob(data)
-		if err != nil {
-			return nil, fmt.Errorf("Index %v: %v", i, err)
-		}
-		serialisedData = append(serialisedData, refinedData...)
-	}
+func isSkipEvm(indicator byte) bool {
```
Comment (prestonvanloon) [inline sharding/utils/marshal.go]: Can you add a comment on how this should work? 
Bitwise operations are not always the most intuitive to read and we will likely have readers wondering how this works.

File: sharding/utils/marshal.go [hunk 7]
```diff
@@ -45,149 +48,140 @@ func ConvertFromRawBlob(blob *RawBlob, i interface{}) error {
...
+	return indicator&skipEvmBits>>7 == 1
+}
+func getDatabyteLength(indicator byte) int {
```
Comment (prestonvanloon) [inline sharding/utils/marshal.go]: Comment on this too please.

File: sharding/collation.go [hunk 8]
```diff
@@ -129,10 +127,7 @@ func ConvertBackToTx(rawBlobs []utils.RawBlob) ([]*types.Transaction, error) {
 
 }
 
-// SerializeTxToBlob method serializes the input tx
-// and returns the blobs in byte array.
-func SerializeTxToBlob(txs []*types.Transaction) ([]byte, error) {
-
+func SerializeTxToRawBlob(txs []*types.Transaction) ([]*utils.RawBlob, error) {
```
[the remark this answers is not in the document]
Comment (rawfalafel) [reply sharding/collation.go]: Ah, yeah, this is a little confusing. This method does RLP encoding, `SerializeTxToBlob` does both RLP encoding and blob serialization. I'll change the names of the methods in this file to make it more consistent and clear.

File: sharding/utils/marshal_test.go (hunk 4 above)
Comment (rawfalafel) [reply sharding/utils/marshal_test.go]: yep!

File: sharding/utils/marshal_test.go (hunk 3 above)
Comment (rawfalafel) [reply sharding/utils/marshal_test.go]: yep, thanks for catching this!

File: sharding/utils/marshal.go (hunk 5 above)
Comment (rawfalafel) [reply sharding/utils/marshal.go]: Ah yeah, this comment's a little out of date. I'll update.

File: sharding/collation_test.go (hunk 1 above)
Comment (rawfalafel) [reply sharding/collation_test.go]: Gotcha. I'll change it to 150 bytes, since that's more in line with what's in the blockchain at the moment. That's also in line with the data that's being encoded in [this post](https://medium.com/@shijuvar/benchmarking-protocol-buffers-json-and-xml-in-go-57fa89b8525).

**Comment by rawfalafel** [comment]
Ready for another look!

**Comment by terenc3t** [comment]
@rawfalafel Lint failed. Don't forget to run gofmt : )

**Comment by rawfalafel** [comment]
Ah, my bad @terenc3t

File: sharding/collation.go [hunk 9]
```diff
@@ -141,28 +121,52 @@ func SerializeTxToBlob(txs []*types.Transaction) ([]byte, error) {
...
 		}
 	}
+	return blobs, nil
+}
+
+// SerializeTxToBlob converts transactions using two steps. First performs RLP encoding, and then blob encoding.
+func SerializeTxToBlob(txs []*types.Transaction) ([]byte, error) {
+	blobs, err := convertTxToRawBlob(txs)
+	if err != nil {
+		return nil, fmt.Errorf("%v", err)
```
[the remark this answers is not in the document]
Comment (rawfalafel) [reply sharding/collation.go]: oh, didn't even notice that!

**Comment by rawfalafel** [comment]
Thanks for the quick turnaround @prestonvanloon !

Btw, I noticed that PRs are getting merged directly into master. Have you discussed squashing the branch into a single commit instead?

**Comment by prestonvanloon** [comment]
@rawfalafel I think we did discuss this. It seems to be the default for our repo now. 

I think this is ready to squash and merge. @nisdas / @terenc3t do you have any other comments?

Thanks so much, Yutaro!

**Comment by nisdas** [comment]
LGTM

**Comment by terenc3t** [comment]
No comment from my end. Thanks so much @rawfalafel
